meta-balena-common/recipes-core/systemd/disk-watchdog/disk-watchdog.c

```diff
@@ -0,0 +1,296 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <getopt.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <systemd/sd-daemon.h>
+
+#define BUFFER_SIZE 512
+#define DEFAULT_INTERVAL 10000  /* 10ms in microseconds */
+
+static volatile int running = 1;
+static char *test_file = NULL;
+static int interval_us = DEFAULT_INTERVAL;
+static int verbose_mode = 0;
+static int debug_mode = 0;
+
+#define LOG_VERBOSE(fmt, ...) do { \
+    if (verbose_mode) printf(fmt, ##__VA_ARGS__); \
+} while(0)
+
+// this test will overwrite the file with test data and check for write errors
+int test_write(const char *filename) {
+    void *write_buf;
+    int fd;
+    ssize_t ret;
+    int i;
+    const int write_chunks = 10; // Write 10 * BUFFER_SIZE
+
+    // it's necessary to align the buffer to the page size because O_DIRECT requires it
+    if (posix_memalign(&write_buf, BUFFER_SIZE, BUFFER_SIZE)) {
+        fprintf(stderr, "posix_memalign failed for write buffer\n");
+        return 1;
+    }
+    memset(write_buf, 'A', BUFFER_SIZE);
+
+    fd = open(filename, O_CREAT | O_WRONLY | O_TRUNC | O_DIRECT, 0666);
+    if (fd < 0) {
+        fprintf(stderr, "open failed: %s\n", strerror(errno));
+        free(write_buf);
+        return 2;
+    }
+
+    // Write multiple chunks to test sustained I/O
+    for (i = 0; i < write_chunks; i++) {
+        ret = write(fd, write_buf, BUFFER_SIZE);
+        if (ret < 0) {
+            fprintf(stderr, "write failed on chunk %d: %s\n", i, strerror(errno));
+            close(fd);
+            free(write_buf);
+            return 3;
+        }
+        if (ret != BUFFER_SIZE) {
+            fprintf(stderr, "partial write on chunk %d: %zd/%d bytes\n", i, ret, BUFFER_SIZE);
+            close(fd);
+            free(write_buf);
+            return 4;
+        }
+    }
+
+    if (fsync(fd) < 0) {
+        fprintf(stderr, "fsync failed: %s\n", strerror(errno));
+        close(fd);
+        free(write_buf);
+        return 5;
+    }
+
+    if (close(fd) < 0) {
+        fprintf(stderr, "close failed: %s\n", strerror(errno));
+        free(write_buf);
+        return 6;
+    }
+
+    free(write_buf);
+
+    sync();
+    sync();
```

Comment:
Why two sync()?

Reply:
Well to be honest I'd say it's overkill, but that's what's watchdog code is doing :

<img width="572" height="354" alt="image" src="https://github.com/user-attachments/assets/9d667cda-2930-43cf-b46b-69b9ba4634d9" />